File: src/squares-of-a-sorted-array/squares-of-a-sorted-array.rs
// https://leetcode.com/problems/squares-of-a-sorted-array

struct Solution;

impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut neg = None;
        let split = nums.partition_point(|&x| x < 0);
        if split > 0 {
            neg = Some(split - 1);
        }
        let mut pos = split;
        let mut out = vec![];

        while neg.is_some() || pos < nums.len() {
            if let Some(n) = neg {
                if pos == nums.len() || -1 * nums[n] < nums[pos] {
                    out.push(nums[n].pow(2));
                    neg = n.checked_sub(1);
                    continue;
                }
            }
            out.push(nums[pos].pow(2));
            pos += 1;
        }

        out
    }
}

fn main() {
    let tests = vec![
        (vec![-4, -1, 0, 3, 10], vec![0, 1, 9, 16, 100]),
        (vec![-7, -3, 2, 3, 11], vec![4, 9, 9, 49, 121]),
        (vec![-4, -2, -1], vec![1, 4, 16]),
        (vec![1, 2, 4], vec![1, 4, 16]),
    ];

    for (nums, solution) in tests {
        let result = Solution::sorted_squares(nums);
        assert_eq!(result, solution);
    }
    println!("✅ All tests passed")
}

File: src/squares-of-a-sorted-array/squares-of-a-sorted-array.py
# https://leetcode.com/problems/squares-of-a-sorted-array

from collections import deque

class Solution(object):
    def sortedSquares(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        pos = deque([])
        neg = deque([])
        out = []
        for num in nums:
            if num < 0:
                neg.append(pow(num, 2))
            else:
                pos.append(pow(num, 2))
        while len(pos) > 0 or len(neg) > 0:
            if len(neg) == 0:
                out.append(pos.popleft())
            elif len(pos) == 0:
                out.append(neg.pop())
            elif neg[-1] < pos[0]:
                out.append(neg.pop())
            else:
                out.append(pos.popleft())
        return out

if __name__ == "__main__":
    sol = Solution()
    tests = [
        ([-4,-1,0,3,10], [0,1,9,16,100]),
        ([-7,-3,2,3,11], [4,9,9,49,121]),
    ]
    for (nums, solution) in tests:
        result = sol.sortedSquares(nums)
        assert result == solution, \
            f"result {result} != solution {solution}"
    print("✅ All tests passed")

